File: code/05-dynamic-programming/c/lcs.c
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Bottom-up, tie-up backtracking. Reused verbatim from
 * lecture-notes/code/lecture05/c/lcs.c (lcs), renamed to match this
 * chapter's pseudocode block name ("LCS Bottom-Up"). Theta(m*n) time/table
 * space. Caller frees *out. */

// snippet:lcs-bottom-up:start
static bool lcs_bottom_up(const char *x, const char *y, char **out) {
    if (!x || !y || !out) return false;
    size_t m=strlen(x),n=strlen(y),w=n+1;
    size_t *dp=calloc((m+1)*w,sizeof(*dp)); if(!dp)return false;
    for(size_t i=1;i<=m;++i)for(size_t j=1;j<=n;++j)
        dp[i*w+j]=(x[i-1]==y[j-1])?dp[(i-1)*w+j-1]+1:
            (dp[(i-1)*w+j]>=dp[i*w+j-1]?dp[(i-1)*w+j]:dp[i*w+j-1]);
    size_t len=dp[m*w+n];char *s=malloc(len+1);if(!s){free(dp);return false;}s[len]='\0';
    size_t i=m,j=n,k=len;
    while(i&&j){if(x[i-1]==y[j-1]){s[--k]=x[i-1];--i;--j;}
        else if(dp[(i-1)*w+j]>=dp[i*w+j-1])--i;else --j;}
    free(dp);*out=s;return true;
}
// snippet:lcs-bottom-up:end

int main(void) {
    const char *x = "ABCBDAB", *y = "BDCABA";
    char *s = NULL;
    lcs_bottom_up(x, y, &s);
    printf("X: %s\n", x);
    printf("Y: %s\n", y);
    printf("lcs: %s\n", s);
    printf("length: %zu\n", strlen(s));
    free(s);
    return 0;
}

File: code/05-dynamic-programming/java/LCS.java
public final class LCS {
    private LCS() {}

    // Bottom-up, tie-up backtracking. Reused verbatim from
    // lecture-notes/code/lecture05/java/LCS.java (solve), renamed to match
    // this chapter's pseudocode block name ("LCS Bottom-Up"). Theta(m*n)
    // time/table space.
    // snippet:lcs-bottom-up:start
    public static String lcsBottomUp(String x, String y) {
        if (x == null || y == null) throw new IllegalArgumentException();
        int m = x.length(), n = y.length();
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (x.charAt(i - 1) == y.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        StringBuilder result = new StringBuilder();
        int i = m, j = n;
        while (i > 0 && j > 0) {
            if (x.charAt(i - 1) == y.charAt(j - 1)) {
                result.append(x.charAt(i - 1));
                i--; j--;
            } else if (dp[i - 1][j] >= dp[i][j - 1]) {
                i--;
            } else {
                j--;
            }
        }
        return result.reverse().toString();
    }
    // snippet:lcs-bottom-up:end

    public static void main(String[] args) {
        String x = "ABCBDAB", y = "BDCABA";
        String s = lcsBottomUp(x, y);
        System.out.println("X: " + x);
        System.out.println("Y: " + y);
        System.out.println("lcs: " + s);
        System.out.println("length: " + s.length());
    }
}
